### PR TITLE
fix: wrong odata count in filter with groupby

### PR DESCRIPTION
**Pull request by larslutz96**
Encase the groupby within an additional SELECT statement to ensure accurate row counting.

### CODE REVIEW

File: db-service/lib/SQLService.js [hunk 1]
```diff
@@ -236,6 +236,18 @@ class SQLService extends DatabaseService {
       expand: false,
       limit: 0,
       orderBy: 0,
+      groupBy: 0,
+      ...(query.SELECT.groupBy
+        ? {
+            from: {
+              SELECT: {
+                columns: [{ ref: [query.SELECT.groupBy[0].ref[0]] }],
```
Comment (BobdenOs) [inline db-service/lib/SQLService.js:244]: This does not look dynamic enough. A GroupBy can have more then 1 column. And the ref can have more then 1 path entry.

Comment (johannes-vogel) [reply db-service/lib/SQLService.js:244]: do we need to specify any column at all?
isn't a select 1 sufficient?

I'd guess that it's sufficient if we ignore any limit clause in the sub select as well.

File: test/scenarios/sflight/read.test.js [hunk 2]
```diff
@@ -44,6 +44,7 @@ describe('SFlight - Read', () => {
     // REVISIT: works in sflight not in tests
     // 'Bookings?$apply=concat(groupby((BookingID,ConnectionID,CurrencyCode_code,FlightDate,ID,TravelID,airline,status))/aggregate($count%20as%20UI5__leaves),aggregate(FlightPrice,CurrencyCode_code),groupby((airline,airlineName),aggregate(FlightPrice,CurrencyCode_code))/concat(aggregate($count%20as%20UI5__count),top(53)))',
     'Bookings?$apply=groupby((airline,airlineName),aggregate(FlightPrice with average as avgPrice,FlightPrice with max as maxPrice,FlightPrice with min as minPrice))&$skip=0&$top=1',
+    `Bookings?$apply=filter(airline eq 'EA' and status eq 'B')/groupby((ID),aggregate(FlightPrice,CurrencyCode_code))&$count=true&$skip=0&$top=1`
```
Comment (johannes-vogel) [inline test/scenarios/sflight/read.test.js:47]: we don't really test that the count is correct.
The request went throught before as well but included a wrong count

File: db-service/lib/SQLService.js (hunk 1 above)
Comment (BobdenOs) [reply db-service/lib/SQLService.js:244]: Wouldn't it make more sense to just wrap the original query (minus limiting factors) with a count.
```cds
SELECT count() FROM (query)
```

File: test/scenarios/sflight/read.test.js (hunk 2 above)
Comment (BobdenOs) [reply test/scenarios/sflight/read.test.js:47]: Also having a more simplified query that reproduces this issue would make it easier to understand the problem later when someone breaks the test.

File: db-service/lib/SQLService.js (hunk 1 above)
Comment (johannes-vogel) [reply db-service/lib/SQLService.js:244]: Yes, I think that would be best.

File: test/scenarios/bookshop/read.test.js [hunk 3]
```diff
@@ -30,6 +30,15 @@ describe('Bookshop - Read', () => {
...
   })
 
+  test('Books $count with $top=2 and filter', async () => {
+    const res = await GET(
+      `/browse/ListOfBooks?$apply=filter(currency_code eq 'GBP')/groupby((ID),aggregate(ID with countdistinct as countBookings))&$count=true&$skip=0&$top=2`,
+    )
+    expect(res.status).to.be.eq(200)
+    expect(res.data.value.length).to.be.eq(2)
+    expect(res.data['@odata.count']).to.be.eq(2)
+  })
```
Comment (BobdenOs) [inline test/scenarios/bookshop/read.test.js:40]:
```suggestion
  test('Books $count with $top=0 and groupby', async () => {
    const res = await GET(
      `/browse/ListOfBooks?$apply=groupby((ID),aggregate(ID with countdistinct as countBookings))&$count=true&$top=2`,
    )
    expect(res.status).to.be.eq(200)
    expect(res.data.value.length).to.be.eq(0)
    expect(res.data['@odata.count']).to.be.eq(5)
  })
```

This change is trying to make the most minimal reproducable scenario. The problem was not the filter it is the `groupby` and the `top` being less then the count.

Comment (larslutz96) [reply test/scenarios/bookshop/read.test.js]: test updated as suggested

**Comment by stewsk** [comment]
Hi @johannes-vogel , would this fix the issue with the Analytical List Page in SFlight?

**Comment by johannes-vogel** [comment]
@stewsk Yes, that's the PR for it.

**Review by BobdenOs — APPROVED** [review]

**Review by johannes-vogel — APPROVED** [review]

**Pull request merged**